src/synth_rapidsilicon.cc

```diff
@@ -8202,6 +8203,101 @@ void collect_clocks (RTLIL::Module* module,
+                }
+            }
+        }
+
+        std::unordered_map<RTLIL::Cell *, int> best_items;
+        for (const auto &cell : dp_items[dsp_limit])
+        {
+            best_items[cell] = items.at(cell); 
+        }
+
+        return {dp[dsp_limit], best_items};
+    }
+
+    void check_dsp_device_limit()
+    {
+        std::unordered_map<RTLIL::Cell *, int> dsp_control;
+
+        for (auto cell : _design->top_module()->cells())
+        {
+            if (cell->type == RTLIL::escape_id("$mul"))
+            {
+                int _a_width_ = cell->getParam(ID::A_WIDTH).as_int();
+                int _b_width_ = cell->getParam(ID::B_WIDTH).as_int();
+                int a_signed = cell->getParam(ID::A_SIGNED).as_int();
+                int b_signed = cell->getParam(ID::B_SIGNED).as_int();
+                if (_a_width_ >= 11 && _b_width_ >= 10)
+                {
+                    if (a_signed && b_signed)
+                    {
+                        double sliceA = static_cast<double>(_a_width_ - a_signed - 1) / (20 - a_signed);
+                        double sliceB = static_cast<double>(_b_width_ - b_signed - 1) / (18 - b_signed);
+                        int nodsp = round(sliceA) * round(sliceB);
+                        dsp_control[cell] = nodsp;
+                    }
+                    else
+                    {
+                        double sliceA = static_cast<double>(_a_width_ - 1) / (20);
+                        double sliceB = static_cast<double>(_b_width_ - 1) / (18);
+                        int nodsp = round(sliceA) * round(sliceB);
+                        dsp_control[cell] = nodsp;
+                    }
+                }
+                else
+                {
+                    dsp_control[cell] = 1;
+                }
+            }
+        }
+        auto result = best_sum_under_dsp_limit(dsp_control, max_dsp - DSP_COUNTER);
+
+        int best_sum = result.first;
+        std::unordered_map<RTLIL::Cell *, int> best_items = result.second;
+        DSP_COUNTER = DSP_COUNTER + best_sum;
+        for (const auto &cell_info : best_items)
+        {
+            cell_info.first->set_bool_attribute(RTLIL::escape_id("valid_map"));
+        }
+        bool dsp_limit_exhausted = false;
+        if (DSP_COUNTER >= max_dsp)
+        {
+            for (auto cell : _design->top_module()->cells())
+            {
+                if (cell->type == RTLIL::escape_id("$mul"))
+                {
+                    if (!cell->get_bool_attribute(RTLIL::escape_id("valid_map")))
+                    {
+                        dsp_limit_exhausted = true;
+                        log_warning("Using soft logic for mult '%s'\n", log_id(cell->name));
+                    }
+                }
+            }
+        }
+        if (dsp_limit_exhausted)
+        {
+            log_warning("DSP exceeds the available DSP block limit on the device; the excess will be mapped to LUTs.\n");
```

Comment:
Please print how many (actual number) DSPs we converted to LUTs. That will help the user select a larger device.